### PR TITLE
DEV: Custom generator for rename site setting migration

### PR DESCRIPTION
**Pull request by lis2**
Ability to automatically generate migration when site setting name is changed. 

Example usage: 

`rails generate site_setting_rename_migration site_description contact_email`

### CODE REVIEW

File: lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    MIGRATION_FILE
+  end
+
+  private
+
+  def validate_setting_name!(name)
+    begin
+      SiteSetting.send(name)
+    rescue NoMethodError
+      say "Site setting with #{name} does not exist"
```
Comment (martin-brennan) [inline lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb:39]: What is `say`? I have never seen it !

File: lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+class SiteSettingRenameMigrationGenerator < Rails::Generators::Base
+  argument :old_name, type: :string, banner: "old setting name", required: true
+  argument :new_name, type: :string, banner: "new setting name", required: true
+
+  class ArgumentError < StandardError
+  end
+
+  def create_migration_file
+    timestamp = Time.zone.now.to_s.tr("^0-9", "")[0..13]
```
Comment (martin-brennan) [inline lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb:11]: It's weird that Rails wouldn't provide a method for this :/

File: lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+class SiteSettingRenameMigrationGenerator < Rails::Generators::Base
+  argument :old_name, type: :string, banner: "old setting name", required: true
+  argument :new_name, type: :string, banner: "new setting name", required: true
+
+  class ArgumentError < StandardError
```
Comment (martin-brennan) [inline lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb:7]: This is unnecessary; `ArgumentError` is already in ruby

File: spec/generator/site_setting_rename_migrations_generator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+require "rails/generators"
+require "generators/site_setting_rename_migration/site_setting_rename_migration_generator"
+
+RSpec.describe SiteSettingRenameMigrationGenerator, type: :generator do
+  it "is generating correct migration" do
```
Comment (martin-brennan) [inline spec/generator/site_setting_rename_migrations_generator_spec.rb:8]:
```suggestion
  it "generates the correct migration" do
```

File: lib/generators/site_setting_rename_migration/USAGE [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Description:
+    Generates migration from old site setting name into new site setting name.
```
Comment (martin-brennan) [inline lib/generators/site_setting_rename_migration/USAGE:2]:
```suggestion
    Generates migration to rename a site setting name from an old name to a new site setting name. This assumes that the new setting is the same data format as the old one.
```

**Review by martin-brennan — APPROVED** [review]
Looks nice, just a few things :)

File: lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb (hunk 1 above)
Comment (lis2) [reply lib/generators/site_setting_rename_migration/site_setting_rename_migration_generator.rb:39]: It is fancy puts :) 

RailsGenerators inherits from Thor::Group
https://api.rubyonrails.org/classes/Rails/Generators/Base.html

And Thor defines it
https://github.com/rails/thor/blob/f1ba900585afecfa13b7fff36968fca3e305c371/lib/thor/shell/basic.rb#L98

File: spec/generator/site_setting_rename_migrations_generator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+require "rails/generators"
+require "generators/site_setting_rename_migration/site_setting_rename_migration_generator"
+
+RSpec.describe SiteSettingRenameMigrationGenerator, type: :generator do
+  it "generates the correct migration" do
+    freeze_time DateTime.parse("2010-01-01 12:00")
+    described_class.start(%w[site_description contact_email], destination_root: "#{Rails.root}/tmp")
+    file_path = "#{Rails.root}/tmp/db/migrate/20100101120000_rename_site_description_setting.rb"
+
+    expected_content = <<-EXPECTED_CONTENT
```
Comment (tgxworld) [inline spec/generator/site_setting_rename_migrations_generator_spec.rb:13]: Are we able to use the `<<~` squiggly heredoc to preserve the indentation of the content in this file?

Comment (lis2) [reply spec/generator/site_setting_rename_migrations_generator_spec.rb]: Good catch, thank you :)

**Pull request merged**